app/models/source_resolver.rb

```diff
@@ -0,0 +1,34 @@
+class SourceResolver
+  def self.resolve(new_source)
+    new(new_source).resolve
+  end
+
+  def initialize(new_source)
+    @new_source = new_source
+  end
+
+  def resolve
+    return if @new_source.save
+    adjust_ranks
+    @new_source.save
+  end
+
+  private
+
+  def adjust_ranks
+    taken_rank_types.each do |rank_type|
+      sources_to_adjust(rank_type).each do |source|
+        source.increment!(rank_type)
+      end
+    end
+  end
+
+  def taken_rank_types
```

Comment:
This is cool how you save, read the errors, and increment. The word `taken` took me a minute to understand. Although I'm not sure there's a better word to describe what's going on. "Occupied" is a synonym but doesn't make sense. 

They're basically conflicts that are being resolved. Maybe there's a better metaphor there... Something like the following:

```ruby
def resolve_ranks
  conflicted_rank_types.each ...
```